When providing a null username or null password Provider error -  panic: runtime error: invalid memory address or nil pointer dereference

### Environment

 * TMOS/Bigip Version: v15.1.5.1
 * Terraform Version: 1.2.8
 * Terraform bigip provider Version: 1.15.1

### Summary

If you provide either a null password or null username the provider panics and produces a stack trace.  I was able to do this with Terraform Cloud where I defined variables in my workspace, but embarrassingly, forgot to assign values to the variables.  I would expect that this exception would be caught and a more useful error message indicating a failure to provide a valid username or password to be created instead.  I think this is a pretty low severity issue, just something that I think might be helpful and save someone else some embarrassing troubleshooting later on.

### Steps To Reproduce

Steps to reproduce the behavior:
This is what I did to reproduce this behavior
1. Create some terraform that creates any resource that requires the bigip provider
2. Create a workspace in Terraform Cloud to run that Terraform plan 
3. Create variables for providing a username and password but then don't assign any value (note if the variable existed before and you erased the value it doesn't seem to reproduce the error)
4. Run the plan and try and apply it.
5. Receive Stack Trace

I have played around with it a bit and this happens if you have both user and password as null, or either one as null.
I can also recreate this using terraform on my workstation by exporting empty username and password environment variables.
`export TF_VAR_username=
export TF_VAR_password=`

And my provider block looks like this
`provider "bigip" {
  address = var.hostname
  username = var.username
  password = var.password
}`

### Expected Behavior
I expect an error message indicating an authorization failure or that no username or password was provided

### Actual Behavior
A stack trace was created like this:
Stack trace from the terraform-provider-bigip_v1.15.1 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x72555f]

goroutine 29 [running]:
github.com/f5devcentral/go-bigip.(*BigIP).APICall(0x0, 0xc0004d55a8)
github.com/f5devcentral/go-bigip@v0.0.0-20220817175609-60a96e5b7c18/bigip.go:177 +0x7f
github.com/f5devcentral/go-bigip.(*BigIP).getForEntity(0x16b2e00?, {0xd99320, 0xc00031cac8}, {0xc0004d5638?, 0xc000134000?, 0xc0004d5610?})
github.com/f5devcentral/go-bigip@v0.0.0-20220817175609-60a96e5b7c18/bigip.go:451 +0xb8
github.com/f5devcentral/go-bigip.(*BigIP).VirtualAddresses(0xc0004d56d0?)
github.com/f5devcentral/go-bigip@v0.0.0-20220817175609-60a96e5b7c18/ltm.go:2541 +0x8c
github.com/F5Networks/terraform-provider-bigip/bigip.resourceBigipLtmVirtualAddressExists(0xc0004c4380, {0xf19b20?, 0x0})
github.com/F5Networks/terraform-provider-bigip/bigip/resource_bigip_ltm_virtual_address.go:156 +0x15a
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc000569280, 0xc000134dc0, {0xf19b20, 0x0})
github.com/hashicorp/terraform-plugin-sdk@v1.1.0/helper/schema/resource.go:440 +0x131
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc00012ebf0, {0xc000134cd0?, 0x51f946?}, 0xc000134cd0)
github.com/hashicorp/terraform-plugin-sdk@v1.1.0/internal/helper/plugin/grpc_provider.go:525 +0x365
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler({0xee1b20?, 0xc00012ebf0}, {0x10df7b0, 0xc0002f9170}, 0xc0004063c0, 0x0)
github.com/hashicorp/terraform-plugin-sdk@v1.1.0/internal/tfplugin5/tfplugin5.pb.go:3153 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001ae480, {0x10e3820, 0xc00024a000}, 0xc000326300, 0xc00008d7a0, 0x16a3370, 0x0)
google.golang.org/grpc@v1.23.0/server.go:995 +0xe1e
google.golang.org/grpc.(*Server).handleStream(0xc0001ae480, {0x10e3820, 0xc00024a000}, 0xc000326300, 0x0)
google.golang.org/grpc@v1.23.0/server.go:1275 +0xa16
google.golang.org/grpc.(*Server).serveStreams.func1.1()
google.golang.org/grpc@v1.23.0/server.go:710 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
google.golang.org/grpc@v1.23.0/server.go:708 +0xea

Error: The terraform-provider-bigip_v1.15.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

--- a/bigip/config.go
+++ b/bigip/config.go
@@ -76,7 +76,7 @@ func (c *Config) Client() (*bigip.BigIP, error) {
 			return client, nil
 		}
 	}
-	return nil, err
+	return client, err
 
 }
 

--- a/bigip/provider.go
+++ b/bigip/provider.go
@@ -197,7 +197,7 @@ func providerConfigure(d *schema.ResourceData, terraformVersion string) (interfa
 		cfg.UserAgent = fmt.Sprintf("Terraform/%s", terraformVersion)
 		cfg.UserAgent += fmt.Sprintf("/terraform-provider-bigip/%s", getVersion())
 		cfg.Teem = d.Get("teem_disable").(bool)
-		// cfg.Transport.TLSClientConfig.InsecureSkipVerify = d.Get("validate_certs_disable").(bool)
+		cfg.Transport.TLSClientConfig.InsecureSkipVerify = d.Get("validate_certs_disable").(bool)
 	}
 	return cfg, err
 }

--- a/bigip/resource_bigip_fast_http.go
+++ b/bigip/resource_bigip_fast_http.go
@@ -68,21 +68,21 @@ func resourceBigipHttpFastApp() *schema.Resource {
 				Type:          schema.TypeString,
 				Optional:      true,
 				Description:   "name of an existing BIG-IP SNAT pool",
-				ConflictsWith: []string{"fast_create_snat_pool_address"},
+				ConflictsWith: []string{"snat_pool_address"},
 			},
-			"fast_create_snat_pool_address": {
+			"snat_pool_address": {
 				Type:          schema.TypeList,
 				Optional:      true,
 				Elem:          &schema.Schema{Type: schema.TypeString},
 				ConflictsWith: []string{"existing_snat_pool"},
 			},
-			"exist_pool_name": {
+			"existing_pool": {
 				Type:          schema.TypeString,
 				Optional:      true,
 				Description:   "Select an existing BIG-IP Pool",
-				ConflictsWith: []string{"fast_create_pool_members"},
+				ConflictsWith: []string{"pool_members"},
 			},
-			"fast_create_pool_members": {
+			"pool_members": {
 				Type:     schema.TypeSet,
 				Optional: true,
 				Elem: &schema.Resource{
@@ -116,7 +116,7 @@ func resourceBigipHttpFastApp() *schema.Resource {
 						},
 					},
 				},
-				ConflictsWith: []string{"exist_pool_name"},
+				ConflictsWith: []string{"existing_pool"},
 			},
 			"load_balancing_mode": {
 				Type:        schema.TypeString,
@@ -152,9 +152,9 @@ func resourceBigipHttpFastApp() *schema.Resource {
 				Type:          schema.TypeString,
 				Optional:      true,
 				Description:   "Select an existing BIG-IP HTTPS pool monitor. Monitors are used to determine the health of the application on each server",
-				ConflictsWith: []string{"exist_pool_name", "fast_create_monitor"},
+				ConflictsWith: []string{"existing_pool", "monitor"},
 			},
-			"fast_create_monitor": {
+			"monitor": {
 				Type:        schema.TypeList,
 				Optional:    true,
 				Description: "foo",
@@ -197,7 +197,7 @@ func resourceBigipHttpFastApp() *schema.Resource {
 						},
 					},
 				},
-				ConflictsWith: []string{"existing_monitor", "exist_pool_name"},
+				ConflictsWith: []string{"existing_monitor", "existing_pool"},
 			},
 		},
 	}
@@ -351,12 +351,12 @@ func setFastHttpData(d *schema.ResourceData, data bigip.FastHttpJson) error {
 	_ = d.Set("slow_ramp_time", data.SlowRampTime)
 	_ = d.Set("monitor.enable", data.MonitorEnable)
 	_ = d.Set("existing_monitor", data.HTTPMonitor)
-	_ = d.Set("fast_create_monitor.0.monitor_auth", data.MonitorAuth)
-	_ = d.Set("fast_create_monitor.0.username", data.MonitorUsername)
-	_ = d.Set("fast_create_monitor.0.password", data.MonitorPassword)
-	_ = d.Set("fast_create_monitor.0.interval", data.MonitorInterval)
-	_ = d.Set("fast_create_monitor.0.send_string", data.MonitorSendString)
-	_ = d.Set("fast_create_monitor.0.response", data.MonitorResponse)
+	_ = d.Set("monitor.0.monitor_auth", data.MonitorAuth)
+	_ = d.Set("monitor.0.username", data.MonitorUsername)
+	_ = d.Set("monitor.0.password", data.MonitorPassword)
+	_ = d.Set("monitor.0.interval", data.MonitorInterval)
+	_ = d.Set("monitor.0.send_string", data.MonitorSendString)
+	_ = d.Set("monitor.0.response", data.MonitorResponse)
 	return nil
 }
 
@@ -418,12 +418,12 @@ func getFastHttpConfig(d *schema.ResourceData) (string, error) {
 	}
 
 	httpJson.PoolEnable = false
-	if v, ok := d.GetOk("exist_pool_name"); ok {
+	if v, ok := d.GetOk("existing_pool"); ok {
 		httpJson.PoolEnable = true
 		httpJson.PoolName = v.(string)
 		httpJson.MakePool = false
 	}
-	if p, ok := d.GetOk("fast_create_pool_members"); ok {
+	if p, ok := d.GetOk("pool_members"); ok {
 		httpJson.PoolEnable = true
 		httpJson.MakePool = true
 		log.Printf("[DEBUG] Adding Pool Members:%+v", p)
@@ -459,7 +459,7 @@ func getFastHttpConfig(d *schema.ResourceData) (string, error) {
 		httpJson.SnatAutomap = false
 		httpJson.MakeSnatPool = false
 	}
-	if s, ok := d.GetOk("fast_create_snat_pool_address"); ok {
+	if s, ok := d.GetOk("snat_pool_address"); ok {
 		httpJson.SnatEnable = true
 		httpJson.SnatAutomap = false
 		httpJson.MakeSnatPool = true
@@ -476,8 +476,8 @@ func getFastHttpConfig(d *schema.ResourceData) (string, error) {
 		httpJson.MonitorEnable = true
 		httpJson.MakeMonitor = false
 	}
-	if s, ok := d.GetOk("fast_create_monitor"); ok {
-		log.Printf("[DEBUG] fast_create_monitor:%+v", s)
+	if s, ok := d.GetOk("monitor"); ok {
+		log.Printf("[DEBUG] monitor:%+v", s)
 		httpJson.MonitorEnable = true
 		httpJson.MakeMonitor = true
 		sL := s.([]interface{})

--- a/bigip/resource_bigip_fast_https.go
+++ b/bigip/resource_bigip_fast_https.go
@@ -75,9 +75,8 @@ func resourceBigipFastHTTPSApp() *schema.Resource {
 				ConflictsWith: []string{"existing_snat_pool"},
 			},
 			"existing_tls_server_profile": {
-				Type:     schema.TypeString,
-				Optional: true,
-				//Default:"/Common/clientssl",
+				Type:          schema.TypeString,
+				Optional:      true,
 				Description:   "Select an existing TLS server profile",
 				ConflictsWith: []string{"tls_server_profile"},
 			},
@@ -102,9 +101,8 @@ func resourceBigipFastHTTPSApp() *schema.Resource {
 				ConflictsWith: []string{"existing_tls_server_profile"},
 			},
 			"existing_tls_client_profile": {
-				Type:     schema.TypeString,
-				Optional: true,
-				//Default:"/Common/clientssl",
+				Type:          schema.TypeString,
+				Optional:      true,
 				Description:   "Select an existing TLS client profile",
 				ConflictsWith: []string{"tls_client_profile"},
 			},
@@ -509,7 +507,7 @@ func getFastHTTPSConfig(d *schema.ResourceData) (string, error) {
 		httpJson.WafPolicyEnable = true
 		httpJson.MakeWafpolicy = true
 		httpJson.AsmLoggingEnable = true
-		//httpJson.WafPolicyName = ""
+		// httpJson.WafPolicyName = ""
 		wafPol := v.([]interface{})
 		for _, vv := range wafPol {
 			log.Printf("[DEBUG] waf_secu policy:%+v", vv)

--- a/bigip/resource_bigip_fast_https.go
+++ b/bigip/resource_bigip_fast_https.go
@@ -66,22 +66,22 @@ func resourceBigipFastHTTPSApp() *schema.Resource {
 				Type:          schema.TypeString,
 				Optional:      true,
 				Description:   "name of an existing BIG-IP SNAT pool",
-				ConflictsWith: []string{"fast_create_snat_pool_address"},
+				ConflictsWith: []string{"snat_pool_address"},
 			},
-			"fast_create_snat_pool_address": {
+			"snat_pool_address": {
 				Type:          schema.TypeList,
 				Optional:      true,
 				Elem:          &schema.Schema{Type: schema.TypeString},
 				ConflictsWith: []string{"existing_snat_pool"},
 			},
-			"tls_server_profile_name": {
+			"existing_tls_server_profile": {
 				Type:     schema.TypeString,
 				Optional: true,
 				//Default:"/Common/clientssl",
 				Description:   "Select an existing TLS server profile",
-				ConflictsWith: []string{"create_tls_server_profile"},
+				ConflictsWith: []string{"tls_server_profile"},
 			},
-			"create_tls_server_profile": {
+			"tls_server_profile": {
 				Type:     schema.TypeList,
 				Optional: true,
 				MaxItems: 1,
@@ -99,15 +99,42 @@ func resourceBigipFastHTTPSApp() *schema.Resource {
 						},
 					},
 				},
-				ConflictsWith: []string{"tls_server_profile_name"},
+				ConflictsWith: []string{"existing_tls_server_profile"},
 			},
-			"exist_pool_name": {
+			"existing_tls_client_profile": {
+				Type:     schema.TypeString,
+				Optional: true,
+				//Default:"/Common/clientssl",
+				Description:   "Select an existing TLS client profile",
+				ConflictsWith: []string{"tls_client_profile"},
+			},
+			"tls_client_profile": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"tls_cert_name": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "Select an existing BIG-IP SSL certificate",
+						},
+						"tls_key_name": {
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "Select an existing BIG-IP SSL key",
+						},
+					},
+				},
+				ConflictsWith: []string{"existing_tls_client_profile"},
+			},
+			"existing_pool": {
 				Type:          schema.TypeString,
 				Optional:      true,
 				Description:   "Select an existing BIG-IP Pool",
-				ConflictsWith: []string{"fast_create_pool_members", "existing_monitor", "fast_create_monitor"},
+				ConflictsWith: []string{"pool_members", "existing_monitor", "monitor"},
 			},
-			"fast_create_pool_members": {
+			"pool_members": {
 				Type:     schema.TypeSet,
 				Optional: true,
 				Elem: &schema.Resource{
@@ -141,7 +168,7 @@ func resourceBigipFastHTTPSApp() *schema.Resource {
 						},
 					},
 				},
-				ConflictsWith: []string{"exist_pool_name"},
+				ConflictsWith: []string{"existing_pool"},
 			},
 			"load_balancing_mode": {
 				Type:        schema.TypeString,
@@ -174,13 +201,33 @@ func resourceBigipFastHTTPSApp() *schema.Resource {
 				Optional:    true,
 				Description: "none",
 			},
+			"existing_waf_security_policy": {
+				Type:          schema.TypeString,
+				Optional:      true,
+				ConflictsWith: []string{"waf_security_policy"},
+			},
+			"waf_security_policy": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enable": {
+							Type:        schema.TypeBool,
+							Required:    true,
+							Description: "Enables Fast to WAF security policy",
+						},
+					},
+				},
+				ConflictsWith: []string{"existing_waf_security_policy"},
+			},
 			"existing_monitor": {
 				Type:          schema.TypeString,
 				Optional:      true,
 				Description:   "Select an existing BIG-IP HTTPS pool monitor. Monitors are used to determine the health of the application on each server",
-				ConflictsWith: []string{"exist_pool_name", "fast_create_monitor"},
+				ConflictsWith: []string{"existing_pool", "monitor"},
 			},
-			"fast_create_monitor": {
+			"monitor": {
 				Type:        schema.TypeList,
 				Optional:    true,
 				Description: "foo",
@@ -222,7 +269,7 @@ func resourceBigipFastHTTPSApp() *schema.Resource {
 						},
 					},
 				},
-				ConflictsWith: []string{"existing_monitor", "exist_pool_name"},
+				ConflictsWith: []string{"existing_monitor", "existing_pool"},
 			},
 		},
 	}
@@ -370,23 +417,32 @@ func setFastHTTPSData(d *schema.ResourceData, data bigip.FastHttpJson) error {
 		return fmt.Errorf("error setting virtual_server: %w", err)
 	}
 	_ = d.Set("existing_snat_pool", data.SnatPoolName)
-	_ = d.Set("fast_create_snat_pool_address", data.SnatAddresses)
-	_ = d.Set("exist_pool_name", data.PoolName)
-	_ = d.Set("tls_server_profile_name", data.TlsServerProfileName)
-	if _, ok := d.GetOk("create_tls_server_profile"); ok {
-		if err := d.Set("create_tls_server_profile", []interface{}{flattenFastTlsServerProfile(data)}); err != nil {
-			return fmt.Errorf("error setting create_tls_server_profile: %w", err)
+	_ = d.Set("snat_pool_address", data.SnatAddresses)
+	_ = d.Set("existing_pool", data.PoolName)
+	_ = d.Set("existing_tls_server_profile", data.TlsServerProfileName)
+	if _, ok := d.GetOk("tls_server_profile"); ok {
+		if err := d.Set("tls_server_profile", []interface{}{flattenFastTlsServerProfile(data)}); err != nil {
+			return fmt.Errorf("error setting tls_server_profile: %w", err)
 		}
+	}
+	_ = d.Set("existing_tls_client_profile", data.TlsClientProfileName)
+	if _, ok := d.GetOk("tls_client_profile"); ok {
+		if err := d.Set("tls_client_profile", []interface{}{flattenFastTlsServerProfile(data)}); err != nil {
+			return fmt.Errorf("error setting tls_client_profile: %w", err)
+		}
+	}
+	if _, ok := d.GetOk("existing_waf_security_policy"); ok {
+		_ = d.Set("existing_waf_security_policy", data.WafPolicyName)
 	}
 	members := flattenFastPoolMembers(data.PoolMembers)
 	log.Printf("[DEBUG]: Pool Members :%+v", members)
-	_ = d.Set("fast_create_pool_members", members)
+	_ = d.Set("pool_members", members)
 	_ = d.Set("load_balancing_mode", data.LoadBalancingMode)
 	_ = d.Set("slow_ramp_time", data.SlowRampTime)
 	_ = d.Set("existing_monitor", data.HTTPMonitor)
-	if _, ok := d.GetOk("fast_create_monitor"); ok {
-		if err := d.Set("fast_create_monitor", []interface{}{flattenFastMonitor(data)}); err != nil {
-			return fmt.Errorf("error setting fast_create_monitor: %w", err)
+	if _, ok := d.GetOk("monitor"); ok {
+		if err := d.Set("monitor", []interface{}{flattenFastMonitor(data)}); err != nil {
+			return fmt.Errorf("error setting monitor: %w", err)
 		}
 	}
 	return nil
@@ -405,6 +461,9 @@ func getFastHTTPSConfig(d *schema.ResourceData) (string, error) {
 		Application: d.Get("application").(string),
 	}
 	httpJson.TlsServerEnable = true
+	httpJson.TlsClientEnable = false
+	httpJson.WafPolicyEnable = false
+	httpJson.AsmLoggingEnable = false
 	httpJson.TlsServerProfileCreate = false
 
 	if v, ok := d.GetOk("virtual_server"); ok {
@@ -414,10 +473,10 @@ func getFastHTTPSConfig(d *schema.ResourceData) (string, error) {
 			httpJson.VirtualPort = v.(map[string]interface{})["port"]
 		}
 	}
-	if v, ok := d.GetOk("tls_server_profile_name"); ok {
+	if v, ok := d.GetOk("existing_tls_server_profile"); ok {
 		httpJson.TlsServerProfileName = v.(string)
 	}
-	if v, ok := d.GetOk("create_tls_server_profile"); ok {
+	if v, ok := d.GetOk("tls_server_profile"); ok {
 		httpJson.TlsServerProfileCreate = true
 		httpJson.TlsServerProfileName = ""
 		tlsSer := v.([]interface{})
@@ -426,13 +485,43 @@ func getFastHTTPSConfig(d *schema.ResourceData) (string, error) {
 			httpJson.TlsKeyName = v.(map[string]interface{})["tls_key_name"].(string)
 		}
 	}
+	if v, ok := d.GetOk("existing_tls_client_profile"); ok {
+		httpJson.TlsClientEnable = true
+		httpJson.TlsClientProfileName = v.(string)
+
+	}
+	if v, ok := d.GetOk("tls_client_profile"); ok {
+		httpJson.TlsClientEnable = true
+		httpJson.TlsClientProfileCreate = true
+		httpJson.TlsClientProfileName = ""
+		tlsClient := v.([]interface{})
+		for _, v := range tlsClient {
+			httpJson.TlsCertName = v.(map[string]interface{})["tls_cert_name"].(string)
+			httpJson.TlsKeyName = v.(map[string]interface{})["tls_key_name"].(string)
+		}
+	}
+	if v, ok := d.GetOk("existing_waf_security_policy"); ok {
+		httpJson.WafPolicyEnable = true
+		httpJson.WafPolicyName = v.(string)
+		httpJson.AsmLoggingEnable = true
+	}
+	if v, ok := d.GetOk("waf_security_policy"); ok {
+		httpJson.WafPolicyEnable = true
+		httpJson.MakeWafpolicy = true
+		httpJson.AsmLoggingEnable = true
+		//httpJson.WafPolicyName = ""
+		wafPol := v.([]interface{})
+		for _, vv := range wafPol {
+			log.Printf("[DEBUG] waf_secu policy:%+v", vv)
+		}
+	}
 	httpJson.PoolEnable = false
-	if v, ok := d.GetOk("exist_pool_name"); ok {
+	if v, ok := d.GetOk("existing_pool"); ok {
 		httpJson.PoolEnable = true
 		httpJson.PoolName = v.(string)
 		httpJson.MakePool = false
 	}
-	if p, ok := d.GetOk("fast_create_pool_members"); ok {
+	if p, ok := d.GetOk("pool_members"); ok {
 		httpJson.PoolEnable = true
 		httpJson.MakePool = true
 		var members []bigip.FastHttpPool
@@ -466,7 +555,7 @@ func getFastHTTPSConfig(d *schema.ResourceData) (string, error) {
 		httpJson.SnatAutomap = false
 		httpJson.MakeSnatPool = false
 	}
-	if s, ok := d.GetOk("fast_create_snat_pool_address"); ok {
+	if s, ok := d.GetOk("snat_pool_address"); ok {
 		httpJson.SnatEnable = true
 		httpJson.SnatAutomap = false
 		httpJson.MakeSnatPool = true
@@ -483,8 +572,8 @@ func getFastHTTPSConfig(d *schema.ResourceData) (string, error) {
 		httpJson.MonitorEnable = true
 		httpJson.MakeMonitor = false
 	}
-	if s, ok := d.GetOk("fast_create_monitor"); ok {
-		log.Printf("[DEBUG] fast_create_monitor:%+v", s)
+	if s, ok := d.GetOk("monitor"); ok {
+		log.Printf("[DEBUG] monitor:%+v", s)
 		httpJson.MonitorEnable = true
 		httpJson.MakeMonitor = true
 		sL := s.([]interface{})

--- a/bigip/resource_bigip_fast_https_test.go
+++ b/bigip/resource_bigip_fast_https_test.go
@@ -135,8 +135,8 @@ resource "bigip_fast_https_app" "fast_https_app" {
     port = 443
   }
   tls_server_profile {
-	tls_cert_name = "/Common/default.crt"
-	tls_key_name = "/Common/default.key"
+    tls_cert_name = "/Common/default.crt"
+    tls_key_name  = "/Common/default.key"
   }
 }
 `, httpsTenantName, httpsAppName)
@@ -152,8 +152,8 @@ resource "bigip_fast_https_app" "fast_https_app" {
     port = 443
   }
   tls_client_profile {
-	tls_cert_name = "/Common/default.crt"
-	tls_key_name = "/Common/default.key"
+    tls_cert_name = "/Common/default.crt"
+    tls_key_name  = "/Common/default.key"
   }
 }
 `, httpsTenantName, httpsAppName)
@@ -169,15 +169,15 @@ resource "bigip_fast_https_app" "fast_https_app" {
     port = 443
   }
   tls_server_profile {
-	tls_cert_name = "/Common/default.crt"
-	tls_key_name = "/Common/default.key"
+    tls_cert_name = "/Common/default.crt"
+    tls_key_name  = "/Common/default.key"
   }
   tls_client_profile {
-	tls_cert_name = "/Common/default.crt"
-	tls_key_name = "/Common/default.key"
+    tls_cert_name = "/Common/default.crt"
+    tls_key_name  = "/Common/default.key"
   }
   waf_security_policy {
-	enable = true
+    enable = true
   }
 }
 `, httpsTenantName, httpsAppName)

--- a/bigip/resource_bigip_fast_https_test.go
+++ b/bigip/resource_bigip_fast_https_test.go
@@ -40,6 +40,78 @@ func TestAccFastHTTPSAppCreateOnBigip(t *testing.T) {
 	})
 }
 
+func TestAccFastHTTPSAppSSLProfileTC1(t *testing.T) {
+	httpsAppName = "fast_https_apptc1"
+	httpsTenantName = "fast_https_tenanttc1"
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAcctPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckFastHTTPSAppDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: getFastHTTPSAppSSLConfig(),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckFastAppExists(httpsAppName, httpsTenantName, true),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "application", "fast_https_apptc1"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "tenant", "fast_https_tenanttc1"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "virtual_server.0.ip", "10.30.40.44"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "virtual_server.0.port", "443"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccFastHTTPSAppSSLProfileTC2(t *testing.T) {
+	httpsAppName = "fast_https_apptc2"
+	httpsTenantName = "fast_https_tenanttc2"
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAcctPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckFastHTTPSAppDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: getFastHTTPSAppSSLConfigTC2(),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckFastAppExists(httpsAppName, httpsTenantName, true),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "application", "fast_https_apptc2"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "tenant", "fast_https_tenanttc2"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "virtual_server.0.ip", "10.30.40.44"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "virtual_server.0.port", "443"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccFastHTTPSAppProfileTC3(t *testing.T) {
+	httpsAppName = "fast_https_apptc3"
+	httpsTenantName = "fast_https_tenanttc3"
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAcctPreCheck(t)
+		},
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckFastHTTPSAppDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: getFastHTTPSAppConfigTC3(),
+				Check: resource.ComposeTestCheckFunc(
+					testCheckFastAppExists(httpsAppName, httpsTenantName, true),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "application", "fast_https_apptc3"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "tenant", "fast_https_tenanttc3"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "virtual_server.0.ip", "10.30.40.45"),
+					resource.TestCheckResourceAttr("bigip_fast_https_app.fast_https_app", "virtual_server.0.port", "443"),
+				),
+			},
+		},
+	})
+}
+
 func getFastHTTPSAppConfig() string {
 	return fmt.Sprintf(`
 resource "bigip_fast_https_app" "fast_https_app" {
@@ -48,6 +120,64 @@ resource "bigip_fast_https_app" "fast_https_app" {
   virtual_server {
     ip   = "10.30.40.44"
     port = 443
+  }
+}
+`, httpsTenantName, httpsAppName)
+}
+
+func getFastHTTPSAppSSLConfig() string {
+	return fmt.Sprintf(`
+resource "bigip_fast_https_app" "fast_https_app" {
+  tenant      = "%v"
+  application = "%v"
+  virtual_server {
+    ip   = "10.30.40.44"
+    port = 443
+  }
+  tls_server_profile {
+	tls_cert_name = "/Common/default.crt"
+	tls_key_name = "/Common/default.key"
+  }
+}
+`, httpsTenantName, httpsAppName)
+}
+
+func getFastHTTPSAppSSLConfigTC2() string {
+	return fmt.Sprintf(`
+resource "bigip_fast_https_app" "fast_https_app" {
+  tenant      = "%v"
+  application = "%v"
+  virtual_server {
+    ip   = "10.30.40.44"
+    port = 443
+  }
+  tls_client_profile {
+	tls_cert_name = "/Common/default.crt"
+	tls_key_name = "/Common/default.key"
+  }
+}
+`, httpsTenantName, httpsAppName)
+}
+
+func getFastHTTPSAppConfigTC3() string {
+	return fmt.Sprintf(`
+resource "bigip_fast_https_app" "fast_https_app" {
+  tenant      = "%v"
+  application = "%v"
+  virtual_server {
+    ip   = "10.30.40.45"
+    port = 443
+  }
+  tls_server_profile {
+	tls_cert_name = "/Common/default.crt"
+	tls_key_name = "/Common/default.key"
+  }
+  tls_client_profile {
+	tls_cert_name = "/Common/default.crt"
+	tls_key_name = "/Common/default.key"
+  }
+  waf_security_policy {
+	enable = true
   }
 }
 `, httpsTenantName, httpsAppName)

--- a/bigip/resource_bigip_ltm_virtual_address.go
+++ b/bigip/resource_bigip_ltm_virtual_address.go
@@ -128,22 +128,22 @@ func resourceBigipLtmVirtualAddressRead(d *schema.ResourceData, meta interface{}
 		return fmt.Errorf("virtual address %s not found", name)
 	}
 	log.Printf("[DEBUG] virtual address configured on bigip is :%v", vas)
-	d.Set("name", name)
+	_ = d.Set("name", name)
 	if err := d.Set("arp", va.ARP); err != nil {
 		return fmt.Errorf("[DEBUG] Error saving ARP to state for Virtual Address  (%s): %s", d.Id(), err)
 	}
-	d.Set("auto_delete", va.AutoDelete)
-	d.Set("conn_limit", va.ConnectionLimit)
-	d.Set("enabled", va.Enabled)
-	d.Set("icmp_echo", va.ICMPEcho)
+	_ = d.Set("auto_delete", va.AutoDelete)
+	_ = d.Set("conn_limit", va.ConnectionLimit)
+	_ = d.Set("enabled", va.Enabled)
+	_ = d.Set("icmp_echo", va.ICMPEcho)
 	if err := d.Set("advertize_route", va.RouteAdvertisement); err != nil {
 		return fmt.Errorf("[DEBUG] Error saving RouteAdvertisement to state for Virtual Address  (%s): %s", d.Id(), err)
 	}
-	traffic_group := va.TrafficGroup
-	if traffic_group == "none" {
-		traffic_group = fmt.Sprintf("/Common/%s", traffic_group)
+	trafficGroup := va.TrafficGroup
+	if trafficGroup == "none" {
+		trafficGroup = fmt.Sprintf("/Common/%s", trafficGroup)
 	}
-	if err := d.Set("traffic_group", traffic_group); err != nil {
+	if err := d.Set("traffic_group", trafficGroup); err != nil {
 		return fmt.Errorf("[DEBUG] Error saving TrafficGroup to state for Virtual Address  (%s): %s", d.Id(), err)
 	}
 

--- a/docs/resources/bigip_fast_http_app.md
+++ b/docs/resources/bigip_fast_http_app.md
@@ -38,11 +38,11 @@ See [virtual server](#virtual-server) below for more details.
 
 * `existing_snat_pool` - (Optional,`string`) Name of an existing BIG-IP SNAT pool.
 
-* `fast_create_snat_pool_address` - (Optional,`list`) List of address to be used for FAST-Generated SNAT Pool.
+* `snat_pool_address` - (Optional,`list`) List of address to be used for FAST-Generated SNAT Pool.
 
 * `exist_pool_name` - (Optional,`string`) Name of an existing BIG-IP pool.
 
-* `fast_create_pool_members` - (Optional,`set`) `fast_create_pool_members` block takes input for FAST-Generated Pool.
+* `pool_members` - (Optional,`set`) `pool_members` block takes input for FAST-Generated Pool.
 See [Pool Members](#pool-members) below for more details.
       
 * `load_balancing_mode` - (Optional,`string`) A `load balancing method` is an algorithm that the BIG-IP system uses to select a pool member for processing a request. F5 recommends the Least Connections load balancing method
@@ -51,7 +51,7 @@ See [Pool Members](#pool-members) below for more details.
                                             
 * `existing_monitor` - (Optional,`string`) Name of an existing BIG-IP HTTPS pool monitor. Monitors are used to determine the health of the application on each server.
 
-* `fast_create_monitor` - (Optional,`set`) `fast_create_monitor` block takes input for FAST-Generated Pool Monitor.
+* `monitor` - (Optional,`set`) `monitor` block takes input for FAST-Generated Pool Monitor.
 See [Pool Monitor](#pool-monitor) below for more details.
 
 ### virtual server
@@ -67,7 +67,7 @@ The `virtual_server` block supports the following:
 
 Using this block will `enable` for FAST-Generated Pool.
 
-The `fast_create_pool_members` block supports the following:
+The `pool_members` block supports the following:
 
 * `addresses` - (Optional , `list`) List of server address to be used for FAST-Generated Pool.
 
@@ -84,7 +84,7 @@ The `fast_create_pool_members` block supports the following:
 
 Using this block will `enable` for FAST-Generated Pool Monitor.
 
-The `fast_create_monitor` block supports the following:
+The `monitor` block supports the following:
 
 * `monitor_auth` - (Optional , `bool`) set `true` if the servers require login credentials for web access on FAST-Generated Pool Monitor. default is `false`.
 

--- a/docs/resources/bigip_fast_https_app.md
+++ b/docs/resources/bigip_fast_https_app.md
@@ -38,16 +38,26 @@ See [virtual server](#virtual-server) below for more details.
 
 * `existing_snat_pool` - (Optional,`string`) Name of an existing BIG-IP SNAT pool.
 
-* `fast_create_snat_pool_address` - (Optional,`list`) List of address to be used for FAST-Generated SNAT Pool.
+* `snat_pool_address` - (Optional,`list`) List of address to be used for FAST-Generated SNAT Pool.
 
-* `tls_server_profile_name` - (Optional,`string`) Name of an existing TLS server profile.
+* `existing_tls_server_profile` - (Optional,`string`) Name of an existing TLS server profile.
 
-* `create_tls_server_profile` - (Optional,`set`) `create_tls_server_profile` block takes input for FAST-Generated TLS Server Profile.
+* `tls_server_profile` - (Optional,`set`) `tls_server_profile` block takes input for FAST-Generated TLS Server Profile.
 See [TLS Server Profile](#tls-server-profile) below for more details.
 
-* `exist_pool_name` - (Optional,`string`) Name of an existing BIG-IP pool.
+* `existing_tls_client_profile` - (Optional,`string`) Name of an existing TLS client profile.
 
-* `fast_create_pool_members` - (Optional,`set`) `fast_create_pool_members` block takes input for FAST-Generated Pool.
+* `tls_client_profile` - (Optional,`set`) `tls_client_profile` block takes input for FAST-Generated TLS client Profile.
+See [TLS Client Profile](#tls-client-profile) below for more details.
+
+* `existing_waf_security_policy` - (Optional,`string`) Name of an existing WAF Security policy.
+
+* `waf_security_policy` - (Optional,`set`) `waf_security_policy` block takes input for FAST-Generated WAF Security Policy.
+See [WAF Security Policy](#waf-security-policy) below for more details.
+
+* `existing_pool` - (Optional,`string`) Name of an existing BIG-IP pool.
+
+* `pool_members` - (Optional,`set`) `pool_members` block takes input for FAST-Generated Pool.
 See [Pool Members](#pool-members) below for more details.
       
 * `load_balancing_mode` - (Optional,`string`) A `load balancing method` is an algorithm that the BIG-IP system uses to select a pool member for processing a request. F5 recommends the Least Connections load balancing method
@@ -56,7 +66,7 @@ See [Pool Members](#pool-members) below for more details.
                                             
 * `existing_monitor` - (Optional,`string`) Name of an existing BIG-IP HTTPS pool monitor. Monitors are used to determine the health of the application on each server.
 
-* `fast_create_monitor` - (Optional,`set`) `fast_create_monitor` block takes input for FAST-Generated Pool Monitor.
+* `monitor` - (Optional,`set`) `monitor` block takes input for FAST-Generated Pool Monitor.
 See [Pool Monitor](#pool-monitor) below for more details.
 
 ### virtual server
@@ -72,17 +82,33 @@ The `virtual_server` block supports the following:
 ### TLS Server Profile
 Using this block will `enable` for FAST-Generated TLS Server Profile.
 
-The `create_tls_server_profile` block supports the following:
+The `tls_server_profile` block supports the following:
 
 * `tls_cert_name` - (Optional , `string`) Name of existing BIG-IP SSL certificate to be used for FAST-Generated TLS Server Profile.
 
 * `tls_key_name` - (Optional , `string`) Name of existing BIG-IP SSL Key to be used for FAST-Generated TLS Server Profile.
 
+### TLS Client Profile
+Using this block will `enable` for FAST-Generated TLS Client Profile.
+
+The `tls_client_profile` block supports the following:
+
+* `tls_cert_name` - (Optional , `string`) Name of existing BIG-IP SSL certificate to be used for FAST-Generated TLS Server Profile.
+
+* `tls_key_name` - (Optional , `string`) Name of existing BIG-IP SSL Key to be used for FAST-Generated TLS Server Profile.
+
+### WAF Security policy
+Using this block will `enable` for FAST-Generated WAF Security Policy
+
+The `waf_security_policy` block supports the following:
+
+* `enable` - (Optional , `bool`) Setting `true` will enable FAST to create WAF Security Policy.
+
 ### Pool Members
 
 Using this block will `enable` for FAST-Generated Pool.
 
-The `fast_create_pool_members` block supports the following:
+The `pool_members` block supports the following:
 
 * `addresses` - (Optional , `list`) List of server address to be used for FAST-Generated Pool.
 
@@ -99,7 +125,7 @@ The `fast_create_pool_members` block supports the following:
 
 Using this block will `enable` for FAST-Generated Pool Monitor.
 
-The `fast_create_monitor` block supports the following:
+The `monitor` block supports the following:
 
 * `monitor_auth` - (Optional , `bool`) set `true` if the servers require login credentials for web access on FAST-Generated Pool Monitor. default is `false`.
 

--- a/docs/resources/bigip_fast_https_app.md
+++ b/docs/resources/bigip_fast_https_app.md
@@ -45,10 +45,14 @@ See [virtual server](#virtual-server) below for more details.
 * `tls_server_profile` - (Optional,`set`) `tls_server_profile` block takes input for FAST-Generated TLS Server Profile.
 See [TLS Server Profile](#tls-server-profile) below for more details.
 
+~> **NOTE** Profile provided by `existing_tls_server_profile` or `tls_server_profile` used for decrypt client-side connections.
+
 * `existing_tls_client_profile` - (Optional,`string`) Name of an existing TLS client profile.
 
 * `tls_client_profile` - (Optional,`set`) `tls_client_profile` block takes input for FAST-Generated TLS client Profile.
 See [TLS Client Profile](#tls-client-profile) below for more details.
+
+~> **NOTE** Profile provided by `existing_tls_client_profile` or `tls_client_profile` used for encrypt server-side connections.
 
 * `existing_waf_security_policy` - (Optional,`string`) Name of an existing WAF Security policy.
 

